### PR TITLE
Opinionated infra pipeline

### PR DESCRIPTION
**Pull request by njrich28**
PR for early review and discussion. The aim of this functionality is to bring all teams pure infra repo pipelines into line with a common pipeline and make it easier to support rollout of deployment target work.

Added `withInfraPipeline` as opinionated pipeline for pure infra repos, eg. shared infra. This new pipeline
 * maps branches to environments in a similar manner to `withPipeline` e.g. `master` -> `aat` + `prod`, `demo` -> `demo`
 * for PRs, runs **plan only** against `aat`
 * supports configuring additional callbacks before and after stages in line with `withPipeline`
 * supports (but not entirely thought about) metrics publishing
 * supports slack notifications

Also
 * added basic unit tests (essentially only compilation tests)
 * added plum-shared-infrastructure to the pipeline tests to integration test the new pipeline
 * metrics publishing works
```[Pipeline] withCredentials
[Pipeline] {
[Pipeline] echo
15:34:28 Publishing Metrics data
[Pipeline] }
```
image

**TODO**
 * add `withParameterizedInfraPipeline` to mirror `withParameterizedPipeline` for sandbox testing
 * docs

### CODE REVIEW

File: vars/sectionInfraBuild.groovy [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+      folderExists('deploymentTarget') {
+        dir('deploymentTarget') {
+          for (int i = 0; i < deploymentTargets.size() ; i++) {
+            spinInfra(product, null, environment, planOnly, subscription, deploymentTargets[i])
+          }
+        }
+      }
+    }
+
+    // TODO get rid of this vault magic?
```
[the remark this answers is not in the document]
Comment (timja) [reply vars/sectionInfraBuild.groovy]: do this now or later?

File: src/uk/gov/hmcts/contino/CommonPipelineDsl.groovy [hunk 2]
```diff
@@ -0,0 +1,29 @@
+package uk.gov.hmcts.contino
+
+abstract class CommonPipelineDsl implements Serializable {
```
Comment (timja) [inline src/uk/gov/hmcts/contino/CommonPipelineDsl.groovy]: what would you think about extracting out the refactoring first and using it in withPipeline?
i.e. this change hasn't been applied to withPipeline, would probably make sense to refactor before this change...

Comment (njrich28) [reply src/uk/gov/hmcts/contino/CommonPipelineDsl.groovy]: yep - kind of did it back-to-front. Started with the intention of refactoring first bet ended up writing the new implementation with the intention of refactoring existing pipeline later. Will do as a separate PR after this change

File: vars/sectionInfraBuild.groovy (hunk 1 above)
[the remark this answers is not in the document]
Comment (njrich28) [reply vars/sectionInfraBuild.groovy]: will do now

**Pull request merged**